### ISSUE TITLE
Support for math operators in expressions

### ISSUE THREAD

**Issue by gregsdennis** [ietf-wg-jsonpath/draft-ietf-jsonpath-base#419]
I think mathematical operators would be a beneficial addition (😏) to the expression syntax.  It would allow things like

```jsonpath
$[?@.a+@.b!=@.c]
```

to check for model consistency.  (Arguably, you just wouldn't serialize `c` as it should be a calculated field, but people do stranger things.)

There are doubtless other use cases.

I have support for this currently in my library.  It's really easy to implement, and I don't think it would be too hard to specify.

I think this is within our charter as @goessner's original implementations supported "underlying scripting language" for expressions, which undoubtedly supported these operators.

**Comment by gregsdennis** [comment]
(I'm happy to defer this until after we've sorted out our function typing issues.)

**Comment by goessner** [comment]
Well ... this might be useful indeed. But implementing arithmethics and specifying it in a clean way are two very different shoes.

We might deal then with:

* `0.1+0.2 == 0.3` problem.
*  Division by Zero.
*  Define EPSILON
* should we allow `+` operator to also concatenate strings ?
* Explicite number type
* rounding
* `sqrt` ... where to stop ?

Alternatively, I can imagine, that a function similar to [CSS calc](https://developer.mozilla.org/en-US/docs/Web/CSS/calc) would be easy to implement and easier to specify.

**Comment by glyn** [comment]
Yes, first class support for mathematical operators will entail a lot of spec work. Function extensions could be used instead.

I suggest we defer this issue and tag it "revisit-after-base-done".

**Comment by gregsdennis** [comment]
The [comparison](https://cburgmer.github.io/json-path-comparison/results/filter_expression_with_addition.html) indicates that many implementations support a path like `$[?(@.key+50==100)]`, but it's split about 50/50 between reading that as

- a math operation: `@.key` + 50
- a `"key+50"` key

I wonder how adding in a couple spaces would do: `$[?(@.key + 50==100)]`.  This _should_ differentiate whether math operations are supported.

**Comment by cabo** [comment]
`member-name-shorthand` cannot contain a `+`, so recognizing `@.key` is not a problem.
(The problem is that adding math adds a ton of additional considerations.  
E.g., what if `@.key` is `"50"` and not `50`, etc.)

**Comment by gregsdennis** [comment]
> member-name-shorthand cannot contain a +

Yeah, it's understood that those implementations aren't spec-compliant.

> The problem is that adding math adds a ton of additional considerations.  E.g., what if @.key is "50" and not 50, etc.

Yeah, it's understood that we'd have to do that stuff.  I don't think we should shy away from it, though.

I still think this is within our charter.

**Comment by gregsdennis** [comment]
We currently forbid `-` in the shorthand name syntax (requiring the brackets syntax instead), so that's not a problem.

**Comment by glyn** [comment]
Deferring until after base done.

**Comment by ohler55** [comment]
If we are considering the ease of use for the end user I would think `$.parts[?(@.x == @y + 3)]` or `$.parts[?(@.x == (@y + 3))]` would be the most natural.

It shouldn't really matter how hard it is to implement if it is better for the end users. Anyone undertaking the task of implement the spec will have to be competent anyway so a little more work shouldn't be that large a hurdle. (IMHO)

**Comment by goessner** [comment]
@ohler55 ... I do understand this very well from a user's point of view. But on the way there will be a lot of spec work to be done. So we are discussing here a way, how functions - in which form - can help to add arithmetic expressions to queries, while having sufficient user acceptance.

I would applaud if some implementers gain experience meanwhile by implementing side by side

* query inline arithmetic.
* encapsulate it in a calc function.

Then they can help to identify edge cases, type collisions and handling of numeric anomalies.

**Comment by ohler55** [comment]
I took the approach described by @danielaparker in OjG but there is no reason all three of the proposed approaches could not be implemented. Having said that, picking one approach as the minimum and offering the others are extensions might be a way to resolve this.

**Comment by goessner** [comment]
Hmm ... as an outcome of this discussion the realisation matures, that inline arithmetic develops as a de-facto standard in current implementations, which is also the natural thing, users expect.

It seems to be best, to defer activities into that direction until after base done, which in fact was the reason, why Glyn closed this issue.

**Comment by gregsdennis** [comment]
I came up with this for basic math support:

```abnf
math-expr = binary-math-expr / unary-math-expr
binary-math-expr = math-operand binary-math-operator math-operand
unary-math-expr = unary-math-operator math-operand
math-operand = number / singular-query / value-function-expr / math-expr / math-group
math-group = "(" math-expr ")"
binary-math-operator = "+" / "-" / "*" / "/"
unary-math-operator = "-"
```

We'd then add `math-expr` as an option on `comparable`

```abnf
comparable = literal / singular-query / value-function-expr / math-expr
```

I believe this gives support for addition, subtraction, multiplication, division, and grouping.

It does allow multiple negations (e.g. `----4`), which is weird.  There's also an ambiguity in `-4` now between

- negative 4 as a `number`
- positive 4 that has been negated

In the end, I'm not sure it makes much of a difference; maybe it saves an operation to have it as "negative 4."

It also doesn't prevent division by zero, but I don't think I've seen a language that does.  Plus we'd have to contend with a path or a function returning zero.  I think the `math-expr` evaluating to `Nothing` is fine.  That would result in a "false" comparison which just wouldn't select the node.

Similarly any path or `ValueType` function which returns a non-number could result in a `Nothing` evaluation as well.

This doesn't support string concatenation (yet).

**Comment by gregsdennis** [comment]
Does the ABNF _need_ to give operator precedence?

`4+5*6` is syntactically valid whether or not the syntax understands that `*` should be performed before `+`.

**Comment by cabo** [comment]
> Does the ABNF _need_ to give operator precedence?

The principle of least surprise says yes: Implementers will expect the AST they derive from the ABNF to be directly useful for a tree interpreter.
